### ISSUE TITLE
mythic_nginx fails to start

### ISSUE THREAD

**Issue by jatiki** [its-a-feature/Mythic#360]
I recentlyupdated my Mythic installation to the latest commit on the master branch. After which nothing was working and the CPU in my VM went to 100% usage permanently. 

After blaming other things I had updated (Kali, MacOS to Sangoma, UTM to. 4.4.4 - Yes I am using a x64 Kali VM on an M2 MacBook Pro) and fiddling with the CPU core and memory settings for the VM I finally found that mythic_nginx was not starting properly.

This is caused by two lines that have been changed in the nginx-docker/config/templates/services.conf.template in commit 3e1c716.

Here is the diff:

-    listen      0.0.0.0:${NGINX_PORT} ${NGINX_USE_SSL};
-    listen      [::]:${NGINX_PORT} ${NGINX_USE_SSL};
+   ${NGINX_USE_IPV4} listen      0.0.0.0:${NGINX_PORT} ${NGINX_USE_SSL};
+    ${NGINX_USE_IPV6} listen      [::]:${NGINX_PORT} ${NGINX_USE_SSL};

These end up in the /etc/nginx/conf.d/services.conf file without the variables being inserted/replaced correctly.

I am no expert in docker-compose, so my current fix was to roll back the change in my local repo and everything is working again. 

I hope someone who knows why the changes were made can come up with a proper fix, to achieve whatever the changes intended to do.

**Comment by its-a-feature** [comment]
Hey! Great question! When you say that the variables weren't inserted correctly, are you able to show what those lines were?  In your `.env` file, those two variables should have been automatically created:
```
NGINX_USE_IPV4="true"
NGINX_USE_IPV6="false"
```
which should have resulted in those getting replaced by either `` or `#` depending on if the value was true or not.

If things aren't getting updated properly, I can work to fix it, I just need to know what you're seeing instead.

For context - those two variables were added to allow people to have more control on which interfaces the nginx container is binding to. Specifically, in some environments IPv4 or IPv6 is disabled on systems which can cause issues. With this, you can opt to listen on IPv4, IPv6, or both.

**Comment by jatiki** [comment]
My services.conf actually started like this:

```
$ head nginx-docker/config/conf.d/services.conf
server {
    ${NGINX_USE_IPV4} listen      0.0.0.0:7443 ssl;
    ${NGINX_USE_IPV6} listen      [::]:7443 ssl;
    ssl_certificate      /etc/ssl/private/mythic-cert.crt;
    ssl_certificate_key  /etc/ssl/private/mythic-ssl.key;
    client_max_body_size 500M; # allows file uploads up to 500 megabytes
    ssl_session_timeout 1d;
    absolute_redirect off;
```

Here are the errror messages from the nginx log:

```
$ sudo ./mythic-cli logs mythic_nginx
/docker-entrypoint.sh: /docker-entrypoint.d/ is not empty, will attempt to perform configuration
/docker-entrypoint.sh: Looking for shell scripts in /docker-entrypoint.d/
/docker-entrypoint.sh: Launching /docker-entrypoint.d/10-listen-on-ipv6-by-default.sh
10-listen-on-ipv6-by-default.sh: info: /etc/nginx/conf.d/default.conf is not a file or does not exist
/docker-entrypoint.sh: Launching /docker-entrypoint.d/20-envsubst-on-templates.sh
20-envsubst-on-templates.sh: Running envsubst on /etc/nginx/templates/services.conf.template to /etc/nginx/conf.d/services.conf
/docker-entrypoint.sh: Launching /docker-entrypoint.d/30-tune-worker-processes.sh
/docker-entrypoint.sh: Configuration complete; ready for start up
2024/01/05 23:30:54 [emerg] 1#1: unknown directive "${NGINX_USE_IPV4}" in /etc/nginx/conf.d/services.conf:2
nginx: [emerg] unknown directive "${NGINX_USE_IPV4}" in /etc/nginx/conf.d/services.conf:2
/docker-entrypoint.sh: /docker-entrypoint.d/ is not empty, will attempt to perform configuration
/docker-entrypoint.sh: Looking for shell scripts in /docker-entrypoint.d/
/docker-entrypoint.sh: Launching /docker-entrypoint.d/10-listen-on-ipv6-by-default.sh
10-listen-on-ipv6-by-default.sh: info: /etc/nginx/conf.d/default.conf is not a file or does not exist
/docker-entrypoint.sh: Launching /docker-entrypoint.d/20-envsubst-on-templates.sh
```

**Comment by jatiki** [comment]
These are the changes I have made to get everything working again:

´´´
$ git diff                            
diff --git a/nginx-docker/config/templates/services.conf.template b/nginx-docker/config/templates/services.conf.template
index ea7f3117..5305f8ce 100644
--- a/nginx-docker/config/templates/services.conf.template
+++ b/nginx-docker/config/templates/services.conf.template
@@ -1,6 +1,8 @@
 server {
-    ${NGINX_USE_IPV4} listen      0.0.0.0:${NGINX_PORT} ${NGINX_USE_SSL};
-    ${NGINX_USE_IPV6} listen      [::]:${NGINX_PORT} ${NGINX_USE_SSL};
+#    ${NGINX_USE_IPV4} listen      0.0.0.0:${NGINX_PORT} ${NGINX_USE_SSL};
+    listen      0.0.0.0:${NGINX_PORT} ${NGINX_USE_SSL};
+#    ${NGINX_USE_IPV6} listen      [::]:${NGINX_PORT} ${NGINX_USE_SSL};
+    listen      [::]:${NGINX_PORT} ${NGINX_USE_SSL};
     ssl_certificate      /etc/ssl/private/mythic-cert.crt;
     ssl_certificate_key  /etc/ssl/private/mythic-ssl.key;
     client_max_body_size 500M; # allows file uploads up to 500 megabytes
´´´

And these are the relevant lines from my .env file:
´´´
NGINX_BIND_LOCALHOST_ONLY="false"
NGINX_HOST="mythic_nginx"
NGINX_PORT="7443"
NGINX_USE_IPV4="true"
NGINX_USE_IPV6="true"
NGINX_USE_SSL="true"
´´´

**Comment by its-a-feature** [comment]
hm... so the values are in your `.env` - i wonder if you could run `sudo ./mythic-cli build mythic_nginx`? My guess is the changes got added, but the container was never re-built to bring the changes in, so it just took things at face value?

**Comment by jatiki** [comment]
I did as you asked and we are back to square one with same errror messages in the nginx log. Here is the result of building and restarting the nginx container:

$ sudo ./mythic-cli start mythic_nginx
[*] No build.env file detected in Mythic's root directory; not supplying build arguments to docker containers
    If you need to supply build arguments to docker containers, create build.env and supply key=value entries there
Stopping mythic_nginx ... 
Stopping mythic_nginx ... done
Removing mythic_nginx ... 
Removing mythic_nginx ... done
Going to remove mythic_nginx
Building mythic_nginx
Sending build context to Docker daemon  55.81kB
Step 1/2 : FROM nginx:1.23-alpine
 ---> 8e75cbc5b25c
Step 2/2 : COPY error403.html /usr/share/nginx/html/
 ---> Using cache
 ---> 93ac9b36c48c
Successfully built 93ac9b36c48c
Successfully tagged mythic_nginx:latest
Creating mythic_nginx ... 
Creating mythic_nginx ... done
[*] Waiting for RabbitMQ to come online (Retry Count = 10)
[*] Attempting to connect to RabbitMQ at 127.0.0.1:5672, attempt 1/10
[+] Successfully connected to RabbitMQ at ***@***.***:5672/mythic_vhost

[*] Waiting for Mythic Server and Nginx to come online (Retry Count = 10)
[*] Attempting to connect to Mythic UI at https://127.0.0.1:7443, attempt 1/10
[-] Failed to make connection to host, retrying in 10s
Get "https://127.0.0.1:7443": dial tcp 127.0.0.1:7443: connect: connection refused
[*] Attempting to connect to Mythic UI at https://127.0.0.1:7443, attempt 2/10
[-] Failed to make connection to host, retrying in 10s
Get "https://127.0.0.1:7443": dial tcp 127.0.0.1:7443: connect: connection refused

**Comment by jatiki** [comment]
From what I understand replacing ${NGINX_USE_IPV4} with "#" or "" should be handled by Mythic_CLI/src/cmd/internal/dockercompose.go similar to ${NGINX_USE_SSL}. Whenever the server.conf is created from the template. 
However I just edited the go script to write "foo" instead of "ssl" and stopped/started the mythic_nginx container (which cause a rebuilt) but the change did not apply. So either this script was not called or the replacing process works entirely differently from what I have guessed.

**Comment by its-a-feature** [comment]
That's exactly what should be happening. When you did an update, did you run `make` again to generate a new `mythic-cli` that leverages the new pieces? If you run `sudo ./mythic-cli version` what do you see?

**Comment by jatiki** [comment]
Yes, I just now came to the same conclusion. I was actually still using and old version of mythic_cli but never noticed. 
```
$ sudo ./mythic-cli version
[*] mythic-cli version:    v0.2.1
[*] Mythic Server version: v3.2.11
[*] React UI Version:      v0.1.48
726195121.254342@[824913216767541] (UTM):* SLSGetNextEventRecordInternal: loc (930.4, 1012.2) conn 0x10e903 LMouseDragged win 0x3de1 (click 1)
```

My fault entirely. Sorry for wasting your time. I thought I had followed the documentation: https://docs.mythic-c2.net/installation/updating-mythic but must have skipped over the part under "Updating patches" and just looked at minor/major release updates. 

Thanks for your support. Your prompt replies were a rather unexpected and pleasant surprise.
